doc/reference/trd-userspace-readable-allow-syscalls.md

```diff
@@ -0,0 +1,183 @@
+from the buffer.
+
+2 System Call API
+=================================
+
+2.1 Userspace Readable Allow (Class ID: 7)
+---------------------------------
+The userspace readable allow syscall follows the same expectations and
+requirements as described for the Read-Write syscall in
+[TRD104 Section 4.4](trd104-syscalls.md#44-read-write-allow-class-id-3), with
+the exception that apps are explicitly allowed to read buffers that have
+been passed to the kernel.
+
+The register arguments for Read-Write Allow system calls are as
+follows. The registers r0-r3 correspond to r0-r3 on CortexM and a0-a3
+on RISC-V.
+
+| Argument         | Register |
+|------------------|----------|
+| Driver number    | r0       |
+| Buffer number    | r1       |
+| Address          | r2       |
+| Size             | r3       |
+
+The Tock kernel MUST check that the passed buffer is contained within
+the calling process's writeable address space. Every byte of a passed
+buffer must be readable and writeable by the process. Zero-length
+buffers may therefore have arbitrary addresses. If the passed buffer is
+not complete within the calling process's writeable address space, the
+kernel MUST return a failure result with an error code of `INVALID`.
+The buffer number specifies which buffer this is. A driver may
+support multiple allowed buffers.
+
+The return variants for Read-Write Allow system calls are `Failure
+with 2 u32` and `Success with 2 u32`.  In both cases, `Argument 0`
+contains an address and `Argument 1` contains a length. When a driver
+implementing the Read-Write Allow system call returns a failure
+result, it MUST return the same address and length as those that were passed
+in the call. When a driver implementing the Read-Write Allow system call
+returns a success result, the returned address and length MUST be those
+that were passed in the previous call, unless this is the first call.
+On the first successful invocation of a particular Read-Write Allow system
+call, an driver implementation MUST return address 0 and size 0.
+
+The syscall class ID is shown below:
+
+| Syscall Class    | Syscall Class Number |
+|------------------|----------------------|
+| Read-Write Allow |           7          |
+
+The standard access model for userspace readable allowed buffers is that userspace can read
+from a buffer while the kernel can read or write. Syncronisation
+methods are required to ensure data consistency but are implementation specifc.
+
+Simultaneous accesses to a buffer from both userspace and the kernel can
+cause issues if not implemented properly. For example a userspace app could
+read partially written data.
+This would result in obscure timing bugs that are hard to detect. Due to this
+each capsule using the userspace readable Allow mechanism MUST document, in a Draft or
+Final Documentary TRD, how consistency of the data exchanged through this
+userspace readable memory region is achieved.
```

Comment:
how it ensures userspace always reads consistent data from a userspace readable buffer.